test/testlib.sh.in

```diff
@@ -1,46 +1,315 @@
+  *) bail "invalid command [[$peer $*]]" ;;
+  esac
+}
+
+# Start the specified tinc daemon.
+# usage: start_tinc { foo | bar | baz }
+start_tinc() {
+  peer=$1
+  shift
+
+  case "$peer" in
+  foo) tinc "$peer" start --logfile="$DIR_FOO/log" -d5 "$@" ;;
+  bar) tinc "$peer" start --logfile="$DIR_BAR/log" -d5 "$@" ;;
+  baz) tinc "$peer" start --logfile="$DIR_BAZ/log" -d5 "$@" ;;
+  *) bail "invalid peer $peer" ;;
+  esac
+}
+
+# Stop all tinc clients.
+stop_all_tincs() {
+  (
+    # In case these pid files are mangled.
+    set +e
+    [ -f "$DIR_FOO/pid" ] && stop_tinc foo
+    [ -f "$DIR_BAR/pid" ] && stop_tinc bar
+    [ -f "$DIR_BAZ/pid" ] && stop_tinc baz
+    true
+  )
+}
+
+# Checks that the number of reachable nodes matches what is expected.
+# usage: require_nodes node_name expected_number
+require_nodes() {
+  stderr "Check that we're able to reach tincd"
+  test "$(tinc "$1" pid | count_lines)" = 1
+
+  stderr "Check the number of reachable nodes for $1 (expecting $2)"
+  actual="$(tinc "$1" dump reachable nodes | count_lines)"
+
+  if [ "$actual" != "$2" ]; then
+    stderr "tinc $1: expected $2 reachable nodes, got $actual"
+    exit 1
+  fi
+}
+
+peer_directory() {
+  case "$peer" in
+  foo) echo "$DIR_FOO" ;;
+  bar) echo "$DIR_BAR" ;;
+  baz) echo "$DIR_BAZ" ;;
+  *) bail "invalid peer $peer" ;;
+  esac
+}
+
+script_status_log() {
+  peer=$1
+  script=$(echo "$2" | sed 's:/:-:g')
+  dir=$(peer_directory "$script")
+  echo "$dir/$$.$peer.$script.script-log"
+}
+
+# Create tincd script which runs asynchronously.
+# Sends SIGCONT to test script when child script finishes executing, or SIGTERM on failure.
+# usage: create_script { foo | bar | baz } { tinc-up | host-down | ... } 'script content'
+create_script() {
+  peer=$1
+  script=$2
+  shift 2
+
+  # This is an append-only log of the script's executions.
+  script_log=$(script_status_log "$peer" "$script")
+  printf '' >"$script_log"
+
+  # This is the line that we should start from when
+  # reading that log while waiting for script events.
+  script_log_line=$script_log.line
+  printf 1 >"$script_log_line"
+
+  dir=$(peer_directory "$script")
+  script=$dir/$script
+
+  cat >"$script" <<EOF
+#!/bin/sh
+set -e
+(
+  (
+    cd "$PWD"
+    scriptname="$scriptname" . ./testlib.sh
+    echo started \$$ "NETNAME=\$NETNAME NAME=\$NAME DEVICE=\$DEVICE INTERFACE=\$INTERFACE NODE=\$NODE REMOTEADDRESS=\$REMOTEADDRESS REMOTEPORT=\$REMOTEPORT SUBNET=\$SUBNET WEIGHT=\$WEIGHT INVITATION_FILE=\$INVITATION_FILE INVITATION_URL=\$INVITATION_URL" >>"$script_log"
+    $@
+    echo finished \$$ >>"$script_log"
+  ) || kill -TERM $$
+) >>'$script.output' 2>&1
+EOF
+
+  chmod u+x "$script"
+
+  if is_windows; then
+    echo "@$MINGW_SHELL '$script'" >"$script.cmd"
+  fi
+}
+
+# Waits for `peer`'s script `script` to finish `count` number of times.
+# usage: wait_script { foo | bar | baz } { tinc-up | host-up | ... } [count=1]
+wait_script() {
```

Comment:
Oof, this doesn't look very pretty. Tinc guarantees that script execution is serialized, so we could just add something like this to those scripts:

```
echo $(($(cat $0.counter) + 1)) >$0.counter
```

And then you can just read the file `script-name.counter` to find out how often the script was called.

Reply:
Sure. The original idea was to add a more general mechanism for waiting on script events (for example, if we require two-way synchronization between the test script and the tinc script, or some kind of a loop in tinc script, or a longer-running tinc script which runs in the background), but I guess it should be added when (if) it is actually needed.

Thanks for taking a look at this. I'll redo and retest everything and ping you back later.

Reply:
Now I remember why it was written that way, including that particular comment about the append-only log.

When a tinc script writes its entry to the log, the test script may already be waiting on it. It can also start waiting after the entry is written (or both can happen at the same time). Thus, we can't simply read the counter after the fact and do the comparison, there has to be a polling loop with a short delay (or tests will run forever), or some kind of a monitoring mechanism.

There's no good solution to the latter that works reliably across all six operating systems these tests run on and does not require [third-party utilities](https://github.com/emcrisostomo/fswatch) which have to be built from source on most platforms.

It comes down to:

1. the current concoction with only adding lines to the log and then monitoring it with `tail -f` and complex [kill-after-timeout](https://github.com/hg/tinc/blob/f7890e11bb04585e15a1599fb51a4efe1cbf5383/test/testlib.sh.in#L250-L256) (which would not be needed if we could require bash and GNU coreutils 8.32+ everywhere thanks to how both bash and that particular version of tail handle SIGPIPE. Having to support other shells, older versions of GNU tail, and BSD tails made simplifying this impossible.).
1. replace script execution log with a simple counter, but also add a polling loop to `wait_script` similar to this:

```sh
# this is a generated tinc script
(
    # ...
    $script_commands
    echo "$(( $(cat "$script_log") + 1 ))" >"$script_log"
) 2>&1 >"$script.output"
```

```sh
wait_script() {
    # ....

    # zero when `wait_script` is called for the first time
    previous_run=$(cat "$script_log".seen)
    try=0

    while [ $try -lt 60 ]; do
        latest_run=$(cat "$script_log")

        if [ "$latest_run" -gt "$previous_run" ]; then
            echo "$latest_run" >"$script_log".seen
            return
        fi

        try=$((try + 1))
    done

    bail "timeout while waiting for script $peer $script"
}
```

The second one is easier to understand and is what I implemented initially now that I think of it. The polling wasn't pretty though, it was producing a lot of noise in the logs (although it can be suppressed), and even with a short wait tests that call `wait_script` frequently (like `algorithms`) were noticeably slower.

I also tried named pipes and dummy processes that are killed inside of tinc scripts and are `wait`ed by the test script; both were very nice for simple one-off cases, but I couldn't make them work reliably when tinc script can be called many times over without even worse spaghetti than is written now.

What solution would you prefer? If it's still the counter and you're fine with a polling loop, I'll put it back, no problem. I can also reduce the current solution to bare essentials, removing everything that's not absolutely required.

Reply:
Ok, I agree that the polling isn't pretty, and `tail -f | grep -m` is a much nicer solution. It's still a bit complicated with keeping track of the last time we waited for the same thing, we could avoid that but then the caller should provide `wait_script` with the total number of matches it should expect in the log file. Let's go with what you currently have.

Reply:
Hm, I see you are also generating a lot more log files. I don't see why we need all the individual `script-log` files, and the `.output` files don't look very helpful. In `scripts.test`, we write one line for every script called to the single file `scripts.out`. Maybe we can use that for all the other tests as well? If we have a single file, then we can also store the line number we want to read from in a shell variable, so no need for `.line` files either.

Reply:
One more thing, I used to be able to run all the tests in parallel with `make -j check`. With your changes, several tests fail. The `algorithms.test` and `security.test` both use port 32755, so that explains those two, but several other fail as well. I haven't looked at the cause of the rest yet, maybe the avoidance of `sleep` statements is bringing more bugs in the test suite to light. But it would be great to have the tests be able to run in parallel.

Reply:
Sure. I combined logs into a single file per each peer. I don't think we can do without the separate counters though. There were some pretty misleading comments there left by myself. Each pair of `$peer` and `$script` needs to know its own location in the file, or they will skip records.

They were moved out of the files and into the dynamically generated variables with such names that they shouldn't collide with anything. Sad that we don't have associative arrays in POSIX shells.

Sorry about the `make -j` thing, I didn't even realize it was supported by the test suite. Fixed.

Reply:
> Sorry about the make -j thing, I didn't even realize it was supported by the test suite. Fixed.

There's still a problem with it, if I run `make -j test` locally, I get this:

```
$ make -j check
...
SKIP: ns-ping.test
PASS: executables.test
PASS: basic.test
[1] + Stopped (tty input)        make -j check
$ 
```

Reply:
Oops. It's yet another of the dozens of these shell-specific things that only manifests in interactive shells and only when `/bin/sh` is dash; not one of the four versions of bash I've tested, not tcsh, not anything else. Fixed.

Tests:

- [Linux, Windows, macOS](https://github.com/hg/tinc/actions/runs/1021624191)
- [FreeBSD](https://builds.sr.ht/~reducer/job/542802)
- [OpenBSD](https://builds.sr.ht/~reducer/job/542803)
- NetBSD package repository is down (again), I can't check it at the moment

We should probably add one more distribution to #280 through containers, something completely different from Ubuntu, like Alpine.